### PR TITLE
Created a custom spending cap component

### LINKED ISSUE #15224: Create a custom spending cap component
## Resources
[Figma Design file](https://www.figma.com/file/n0xMnZqQz4w9Sfa3ZdhWUi/Consentful-experience-for-token-allowances?node-id=910%3A1321)

## Acceptance Criteria

1. It follows the input logic described in the Figma file

Screen Shot 2022-07-13 at 15 40 44

### PR DESCRIPTION
**Pull request by VSaric**
## Explanation

Created a custom spending cup component. This component can't yet be added to the app. Instead, I added a storybook story that displays this component. This component received at least the following props:

- tokenName (type: string)
- currentTokenBalance (type: number)
- dappProposedValue (type: number)
- siteOrigin (type: string)
- onEdit (type: function)

## More Information

* Fixes #15224 

## Screenshots/Screencaps

https://user-images.githubusercontent.com/92527393/192730850-1d116e39-e289-4682-9cf3-f6ee450c47ae.mov

## Manual Testing Steps

This component can be tested by running storybook (`yarn storybook`) and then go to the `COMPONENTS` tab and open in `App/CustomSpendingCup`. For testing this component you can see the Acceptance Criteria in the ticket: #15224.

### CODE REVIEW

File: ui/components/ui/form-field/form-field.js
Comment (brad-decker) [inline ui/components/ui/form-field/form-field.js:87]: Okay Yikes. After reading this I am gonna have to send this back to the drawing board in terms of composability and design system. What you've done here is add a *second set of tooltip props that can't be combined with the others* If someone were to provide both tooltipContentText and tooltipText it'd result in TWO icons with tooltips.

@georgewrmarshall can I get your opinion: Should we convert InfoTooltip to 'IconWithTooltip" and use a level "warning | info | error" to change the icon and color?  Or create a new component in the design system that serves this purpose.

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: ui/components/ui/custom-spending-cup/custom-spending-cup.js
Comment (VSaric) [reply ui/components/ui/custom-spending-cup/custom-spending-cup.js]: @brad-decker Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/94529dfd5cbf3a62b47e728eec565a0a71dde769).

File: ui/components/ui/custom-spending-cup/index.scss
Comment (VSaric) [reply ui/components/ui/custom-spending-cup/index.scss]: @brad-decker Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/94529dfd5cbf3a62b47e728eec565a0a71dde769).

File: ui/components/ui/custom-spending-cup/custom-spending-cup.js
Comment (VSaric) [reply ui/components/ui/custom-spending-cup/custom-spending-cup.js]: @brad-decker I forgot to remove color error default. I used `WARNING_DEFAULT` and removed the extra css rule. Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/94529dfd5cbf3a62b47e728eec565a0a71dde769#diff-c06a267665dad89208a2494f16f61d99163bbedddbbdce562e911be7077b4676).

File: ui/components/ui/form-field/form-field.js
Comment (VSaric) [reply ui/components/ui/form-field/form-field.js]: @brad-decker Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/94529dfd5cbf3a62b47e728eec565a0a71dde769#diff-c06a267665dad89208a2494f16f61d99163bbedddbbdce562e911be7077b4676).

File: ui/components/ui/custom-spending-cup/custom-spending-cup.js
Comment (VSaric) [reply ui/components/ui/custom-spending-cup/custom-spending-cup.js]: @brad-decker Yes, these paddings are for different screen sizes. I saw in #15106 how it is used and I apply it like that. Thank you for sharing this! Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/94529dfd5cbf3a62b47e728eec565a0a71dde769#diff-c06a267665dad89208a2494f16f61d99163bbedddbbdce562e911be7077b4676).

File: ui/components/ui/form-field/form-field.js
Comment (VSaric) [reply ui/components/ui/form-field/form-field.js]: @brad-decker Yes, I know that this is written ugly and it yikes, but I saw that the property `tooltipText` have only `InfoTooltip`. I must have warning and question icon with tooltip. Base on that, I added this like it is now. I knew it is not the best approach at all. I will wait to @georgewrmarshall answer on this and I will made appropriate change.

Comment (brad-decker) [reply ui/components/ui/form-field/form-field.js]: @VSaric my comment wasn't meant to be derogatory towards your work, I meant that I could see the predicament you are in and we need to think about the solution a bit more to provide you some guidance. I apologize for not being clearer with my intentions in the original comment :(.

Comment (VSaric) [reply ui/components/ui/form-field/form-field.js]: @georgewrmarshall Can we get your opinion on this?

File: app/_locales/en/messages.json
Comment (georgewrmarshall) [inline app/_locales/en/messages.json:1660]: @coreyjanssen is our content guideline to remove trailing periods from the ends of content?

```suggestion
    "message": "Only enter a number that you're comfortable with the contract spending now or in the future. You can always increase the spending cap later"
```

File: ui/components/ui/custom-spending-cap/custom-spending-cap.js
Comment (georgewrmarshall) [inline ui/components/ui/custom-spending-cap/custom-spending-cap.js:105]: Can this not be added to the padding props in the parent `Box` component above?

File: ui/components/ui/form-field/form-field.js
Comment (georgewrmarshall) [inline ui/components/ui/form-field/form-field.js:106]: Another pattern we could add to this component is prop objects for each of the individual components so they are easy to change in their use case

```suggestion
              marginBottom={3}
              marginRight={2}
              {...titleDetailProps}
```

**Review by georgewrmarshall — COMMENTED** [review]
Sorry for the late review. Have made some suggestions. I highly recommend reading the inversion of control article by Kent C Dodds. I also think this should live in `ui/components/app` as it's quite a specific component unlike other `ui/components/ui` components such as `Button` or `FormField`

File: ui/components/ui/form-field/form-field.js
Comment (VSaric) [reply ui/components/ui/form-field/form-field.js]: @georgewrmarshall Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/190f93d7a3b816927c6d40f7039080fdc87be55b).

File: ui/components/ui/custom-spending-cap/custom-spending-cap.js
Comment (VSaric) [reply ui/components/ui/custom-spending-cap/custom-spending-cap.js]: @georgewrmarshall Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/190f93d7a3b816927c6d40f7039080fdc87be55b).

File: ui/components/ui/custom-spending-cap/index.scss
Comment (VSaric) [reply ui/components/ui/custom-spending-cap/index.scss]: @georgewrmarshall Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/190f93d7a3b816927c6d40f7039080fdc87be55b).

File: ui/components/ui/custom-spending-cap/index.scss
Comment (VSaric) [reply ui/components/ui/custom-spending-cap/index.scss]: @georgewrmarshall Yes, I wrapped div in storybook and set the width there to display how it might be in the app. I removed that width's and the parent container is now responsible. Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/190f93d7a3b816927c6d40f7039080fdc87be55b).

File: app/_locales/en/messages.json
Comment (VSaric) [reply app/_locales/en/messages.json:4222]: @georgewrmarshall We are waiting @coreyjanssen to answer on this.

File: app/_locales/en/messages.json
Comment (VSaric) [reply app/_locales/en/messages.json:1667]: @georgewrmarshall We are waiting @coreyjanssen to answer on this.

File: app/_locales/en/messages.json
Comment (VSaric) [reply app/_locales/en/messages.json:1660]: @georgewrmarshall We are waiting @coreyjanssen to answer on this.

File: ui/components/ui/form-field/form-field.js
Comment (VSaric) [reply ui/components/ui/form-field/form-field.js]: Thank you @georgewrmarshall for answer on this! I added a `custom-spending-cup-tooltip.js` file and imported in `custom-spending-cup.js`. Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/190f93d7a3b816927c6d40f7039080fdc87be55b). Let me know if it is okay.
Thank you in advance! 😄

Comment (brad-decker) [reply ui/components/ui/form-field/form-field.js]: great suggestion @georgewrmarshall and @VSaric great work on the cleanup. It already looks 1000% better. I may not get time to review fully today but if i have time i will. If not, Monday morning i will review fully :)

**Review by georgewrmarshall — COMMENTED** [review]
I think we can still extend the `FormField` component a bit more I've created this draft PR https://github.com/MetaMask/metamask-extension/pull/15679 if these updates were made would you be able to use the `FormField` without needing to add any context specific props like `customTooltipComponent`?

File: app/_locales/en/messages.json
Comment (coreyjanssen) [reply app/_locales/en/messages.json:1660]: Just saw this, apologies for the delay! Let's keep the period, as the guideline is to use a period when the content is 2 or more sentences. Thanks!

Comment (georgewrmarshall) [reply app/_locales/en/messages.json:1660]: Ah my mistake sorry @VSaric! 🙇 Noted for future reviews. Thanks for clarifying @coreyjanssen 🙏

File: app/_locales/en/messages.json
Comment (georgewrmarshall) [reply app/_locales/en/messages.json:1667]: Clarified in above reply. 
> When there is more than one sentence, all sentences should end with a period.

File: app/_locales/en/messages.json
Comment (georgewrmarshall) [reply app/_locales/en/messages.json:4222]: Clarified in previous comment
> When there is more than one sentence, all sentences should end with a period.

**Comment by VSaric** [comment]
This PR is blocked by: #15679 @georgewrmarshall

**Comment by brad-decker** [comment]
@VSaric I believe this is unblocked now. mind getting this ready for another review?

**Comment by VSaric** [comment]
> I think we can still extend the `FormField` component a bit more I've created this draft PR #15679 if these updates were made would you be able to use the `FormField` without needing to add any context specific props like `customTooltipComponent`?

I resolved conflicts in `form-field.js` component and used properties that you told me here: https://github.com/MetaMask/metamask-extension/pull/15679#discussion_r963838251. Now it is ready for another review! 😃  @georgewrmarshall  @brad-decker

**Comment by mirjanaKukic** [comment]
Verified by QA

**Review by brad-decker — DISMISSED** [review]
LGTM

File: ui/components/ui/form-field/form-field.js
Comment (georgewrmarshall) [inline ui/components/ui/form-field/form-field.js:189]: We are still adding feature specific content to the ui components? Can this be added in the spending cap component?

**Review by georgewrmarshall — COMMENTED** [review]
Looks like there is still feature specific updates going into the form-field component? All the form-field updates should be made in the spending cup components. Or the form-field should be refactored to allow for any components not just a max button

File: ui/components/ui/form-field/form-field.js
Comment (brad-decker) [reply ui/components/ui/form-field/form-field.js:189]: I missed this in my review, definitely don't need to be passing props to let generic components be aware of their higher level consumer's contexts like this.

File: ui/components/app/custom-spending-cap/custom-spending-cap.js
Comment (VSaric) [reply ui/components/app/custom-spending-cap/custom-spending-cap.js]: Sorry for late fixing this @georgewrmarshall, I was on vacation. Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/1de7f03ba5fa31864eeaf3d69bf7ae9be95d603f).

File: ui/components/ui/form-field/form-field.js
Comment (VSaric) [reply ui/components/ui/form-field/form-field.js]: Yes, I added this in the spending cap component. Sorry for late fixing @georgewrmarshall and @brad-decker, I was on vacation. Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/1de7f03ba5fa31864eeaf3d69bf7ae9be95d603f).

**Review by brad-decker — DISMISSED** [review]

File: ui/components/ui/form-field/form-field.js
Comment (brad-decker) [reply ui/components/ui/form-field/form-field.js]: Hope your vacation was restorative @VSaric , glad to have you back!

**Comment by bschorchit** [comment]
@georgewrmarshall when you have time, could you re-review this PR? Thanks! 🧡

File: ui/components/ui/form-field/form-field.js
Comment (darkwing) [inline ui/components/ui/form-field/form-field.js:120]: Nit: We've been using the America-centric `color` instead

**Review by darkwing — CHANGES_REQUESTED** [review]
One super tiny nit, otherwise good!

**Review by darkwing — DISMISSED** [review]
Fixed the color spelling :)

File: ui/components/ui/form-field/form-field.js
Comment (georgewrmarshall) [inline ui/components/ui/form-field/form-field.js:300]: These also seem context specific to the max cup spending component? `maxButton` could be renamed to something like `rightAccessory`. What does the `coloredValue` indicate? Is it always the same color? Should this be added as a conditional class to the formfield in the spending cup component?

**Review by georgewrmarshall — COMMENTED** [review]
Thanks for making those updates! Happy to proceed to unblock but I still think there are some improvements that can be made. I'm also working on the new TextInput component so it could be that the FormField eventually gets replaced.

File: ui/components/ui/form-field/form-field.js
Comment (VSaric) [reply ui/components/ui/form-field/form-field.js]: Thank you @darkwing for fixing this color spelling! I appreciate it! 😄

File: ui/components/ui/form-field/form-field.js
Comment (VSaric) [reply ui/components/ui/form-field/form-field.js]: @georgewrmarshall You are right, I deleted context specific updates from `form-field` component and added in the spending cap component. [Here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/a70f94956c880932ea53f17e9f54a7fe1641933f) are those changes that I made. 

It will be good to re-review this PR @georgewrmarshall, if you have time, because this PR is a blocker for this ticket: #15211.  😄 

Also @bschorchit comment [here](https://github.com/MetaMask/metamask-extension/issues/15367#issuecomment-1255392751) that this is a stretch goal to be merged to unblocked that ticket and to try to merge all this tickets by September 29th so to gets in the 10.20 release that will be out early in October.

I updated the `Error case` for custom spending cap component also in the latest commit @bschorchit. 😄

**Comment by VSaric** [comment]
@georgewrmarshall @brad-decker @darkwing When you have time, could you re-review this PR? 
Thank you in advance! 😄

File: ui/components/app/custom-spending-cap/index.scss
Comment (georgewrmarshall) [inline ui/components/app/custom-spending-cap/index.scss:17]: I think you want absolute positioning here. Also will need relative positioning for the wrapping container

File: ui/components/ui/form-field/form-field.js
Comment (georgewrmarshall) [inline ui/components/ui/form-field/form-field.js:213]: Oh thanks for the fix 🙏

**Review by georgewrmarshall — COMMENTED** [review]
Looking really good! One little fix to the positioning of the max button otherwise looks g2g!!

File: ui/components/app/custom-spending-cap/index.scss
Comment (VSaric) [reply ui/components/app/custom-spending-cap/index.scss]: @georgewrmarshall Done [here](https://github.com/MetaMask/metamask-extension/pull/15522/commits/e029de4c594789c4ca7605e0485f4493204b8c87).

**Review by brad-decker — APPROVED** [review]

**Review by georgewrmarshall — APPROVED** [review]
LGTM! Great work 💯

**Pull request merged**
